src/wpublish/QueriesAndChunks.cc

```diff
@@ -381,31 +357,57 @@ void QueriesAndChunks::examineAll() {
             string const& slowestTable = begin->db + ":" + begin->table;
             auto iterTbl = scanTblSums.find(slowestTable);
             if (iterTbl != scanTblSums.end()) {
-                LOGS(_log, LOG_LVL_DEBUG, "examineAll " << slowestTable << " chunkId=" << task->getChunkId());
+                LOGS(_log, LOG_LVL_DEBUG,
+                     "examineAll a " << slowestTable << " chunkId=" << task->getChunkId() << " "
+                                     << task->getIdStr());
                 ScanTableSums& tblSums = iterTbl->second;
+
+                // TODO: The size and expected percentage of chunks needs to be gathered from
+                // replicator data and/or the czar.
+                // This code pre-dates the replicator, so it constructs
+                // the tables as they are used by tasks. This causes serious issues
+                // with analyzing queries the first time the table is seen. Skipping analysis
+                // for slow tasks on large tables is particularly brutal, so this assumes
+                // that the table for the first chunk in the list is a close enough
+                // approximation for the percentage.
+                // Use the correct chunk id, if it can be found. Otherwise, use the first
+                // chunk in the list as a reference. Use whatever value is found.
                 auto iterChunk = tblSums.chunkPercentages.find(task->getChunkId());
+                if (iterChunk == tblSums.chunkPercentages.end()) {
+                    iterChunk = tblSums.chunkPercentages.begin();
+                }
                 if (iterChunk != tblSums.chunkPercentages.end()) {
+                    LOGS(_log, LOG_LVL_DEBUG,
+                         "examineAll c " << slowestTable << " chunkId=" << task->getChunkId() << " "
+                                         << task->getIdStr());
                     // We can only make the check if there's data on past chunks/tables.
                     double percent = iterChunk->second.percent;
                     bool valid = iterChunk->second.valid;
-                    double maxTimeChunk = percent * schedMaxTime;
+                    // Allow for some variation in time, also need twice as much time for 2 replicas.
+                    double leeway = 2.25;
```

Comment:
What this magic number is about? And why `2.25`? What if there would be 3 replicas or 1?

Reply:
In the near future, the actual number of chunks in the full table scan, all the sizes of all the tables, and the number of replicas will be available, and the code will use those. This magic number will go away, and hopefully be an unhappy memory. 

`leeway` is meant to be large enough to keep from kicking innocent queries. As the TODO statement above tries to explain, there is no good value for this. The largest chunks are a lot larger than the smallest chunks, and there's no way for the worker to know how many chunks are in a real full table scan, so it's guessing, and it's probably not a even a very good guess. It really needs a reliable source of truth, which should be available soon. Even on the same system, the chunks used on a full tables scan may change at the whim of xrootd, and the worker has no way to know.

See [DM-43131](https://rubinobs.atlassian.net/browse/DM-43131) A reliable source for the details makes this problem a lot more tractable.

Reply:
> In the near future, the actual number of chunks in the full table scan, all the sizes of all the tables, and the number of replicas will be available, and the code will use those. This magic number will go away, and hopefully be an unhappy memory.

This is a wrong design assumption since workers should never have access to the number of replicas. That would require workers to effectively know what chunks other workers have. Implementing that would be a total mess for the Replication system due to complex state transitions. The Replication system would need to be made significantly more complex to redistribute all that information across workers. Besides, in the "shared nothing" architecture of Qserv workers are supported to act independently of each other.

If it's one of the key elements of the "booting" algorithm then, please, reconsider it.

The only option that I can see here would be if you implemented your own Czar-workers protocol where you could send this information from Czar to workers in each query as needed. Knowing table sizes is still fine.

Reply:
What the worker really needs to make a reasonable estimate is the list of chunks it will be expectected to handle in a full tables scan and the size of the tables in those chunks. This will probably come from the czar. 

Replication factor is only useful if the chunks in a full table scan is not available, as a replication factor of 2 would have half the chunks on the worker in a full table scan, a replication factor of 3 would use a third, etc. But, it's much better to just get the actual list of chunks from the czar.